src/openrct2/actions/RideCreateAction.hpp

```diff
@@ -0,0 +1,314 @@
+{
+    RideCreateGameActionResult() : GameActionResult(GA_ERROR::OK, 0) {}
+    RideCreateGameActionResult(GA_ERROR error, rct_string_id message) : GameActionResult(error, message) {}
+
+    sint32 rideIndex;
+    uint32 rideColor;
+};
+
+struct RideCreateAction : public GameActionBase<GAME_COMMAND_CREATE_RIDE, GA_FLAGS::ALLOW_WHILE_PAUSED, RideCreateGameActionResult>
+{
+public:
+    sint32 rideType;
+    sint32 rideSubType;
+
+    void Serialise(DataSerialiser& stream) override
+    {
+        GameAction::Serialise(stream);
+
+        stream << rideType << rideSubType;
+    }
+
+    GameActionResult::Ptr Query() const override
+    {
+        if (rideType >= RIDE_TYPE_COUNT)
+        {
+            return std::make_unique<RideCreateGameActionResult>(GA_ERROR::INVALID_PARAMETERS, STR_INVALID_RIDE_TYPE);
+        }
+
+        return std::make_unique<RideCreateGameActionResult>();
+    }
+
+    GameActionResult::Ptr Execute() const override
+    {
+        rct_ride_entry * rideEntry;
+        auto res = std::make_unique<RideCreateGameActionResult>();
+
+        sint32 subType = rideSubType;
+        if (subType == RIDE_ENTRY_INDEX_NULL)
+        {
+            uint8 *availableRideEntries = get_ride_entry_indices_for_ride_type(rideType);
+            for (uint8 *rei = availableRideEntries; *rei != RIDE_ENTRY_INDEX_NULL; rei++)
+            {
+                rideEntry = get_ride_entry(*rei);
+
+                // Can happen in select-by-track-type mode
+                if (!ride_entry_is_invented(*rei) && !gCheatsIgnoreResearchStatus)
+                {
+                    continue;
+                }
+
+                if (!(rideEntry->flags & RIDE_ENTRY_FLAG_SEPARATE_RIDE_NAME) || rideTypeShouldLoseSeparateFlag(rideEntry))
+                {
+                    subType = *rei;
+                    break;
+                }
+            }
+            if (subType == RIDE_ENTRY_INDEX_NULL)
+            {
+                subType = availableRideEntries[0];
+                if (subType == RIDE_ENTRY_INDEX_NULL)
+                {
+                    res->Error = GA_ERROR::INVALID_PARAMETERS;
+                    res->ErrorMessage = STR_INVALID_RIDE_TYPE;
+                    return std::move(res);
+                }
+            }
+        }
+
+        if (subType >= 128)
+        {
+            res->Error = GA_ERROR::INVALID_PARAMETERS;
+            res->ErrorMessage = STR_INVALID_RIDE_TYPE;
+            return std::move(res);
+        }
+
+        sint32 rideEntryIndex = subType;
+        sint32 rideIndex = ride_get_empty_slot();
+        if (rideIndex == -1) 
+        {
+            res->Error = GA_ERROR::DISALLOWED;
+            res->ErrorMessage = STR_TOO_MANY_RIDES;
+            return std::move(res);
+        }
+
+        res->rideIndex = rideIndex;
+        res->rideColor = ride_get_random_colour_preset_index(rideType) | (ride_get_unused_preset_vehicle_colour(rideType, subType) << 8);
```

Comment:
How will the client and server have matching colours when the server and client call this `util_rand`. Previously the `Query` decided the ride colour on the server and the server would pass out a game command with the colour assigned. Now I'm not sure how this would work.